hy/importer.py

```diff
@@ -163,63 +172,117 @@ def write_hy_as_pyc(fname):
+                                             'exec', flags)
 
-        ispkg = self.is_package(fullname)
 
-        mod.__file__ = self.path
-        mod.__loader__ = self
-        mod.__name__ = fullname
-
-        if ispkg:
-            mod.__path__ = []
-            mod.__package__ = fullname
-        else:
-            mod.__package__ = fullname.rpartition('.')[0]
+# Provide a working namespace for our new FileFinder.
+class HyFileFinder(FileFinder):
+    pass
 
-        sys.modules[fullname] = mod
-        return mod
 
+# Taken from inspect.py and modified to support Hy suffixes.
+def getmodulename(path):
+    fname = os.path.basename(path)
+    suffixes = [(-len(suffix), suffix)
+                for suffix in all_suffixes() + HY_SOURCE_SUFFIXES]
+    suffixes.sort()  # try longest suffixes first, in case they overlap
+    for neglen, suffix in suffixes:
+        if fname.endswith(suffix):
+            return fname[:neglen]
+    return None
 
-class MetaImporter(object):
-    def find_on_path(self, fullname):
-        fls = ["%s/__init__.hy", "%s.hy"]
-        dirpath = "/".join(fullname.split("."))
 
-        for pth in sys.path:
-            pth = os.path.abspath(pth)
-            for fp in fls:
-                composed_path = fp % ("%s/%s" % (pth, dirpath))
-                if os.path.exists(composed_path):
-                    return composed_path
+# Taken from pkgutil.py and modified to support Hy suffixes.
+def _iter_hy_file_finder_modules(importer, prefix=''):
+    if importer.path is None or not os.path.isdir(importer.path):
+        return
 
-    def find_module(self, fullname, path=None):
-        path = self.find_on_path(fullname)
-        if path:
-            return MetaLoader(path)
-
-
-sys.meta_path.insert(0, MetaImporter())
+    yielded = {}
+    try:
+        filenames = os.listdir(importer.path)
+    except OSError:
+        # ignore unreadable directories like import does
+        filenames = []
+    filenames.sort()  # handle packages before same-named modules
+
+    for fn in filenames:
+        modname = getmodulename(fn)
+        if modname == '__init__' or modname in yielded:
+            continue
+
+        path = os.path.join(importer.path, fn)
+        ispkg = False
+
+        if not modname and os.path.isdir(path) and '.' not in fn:
+            modname = fn
+            try:
+                dircontents = os.listdir(path)
+            except OSError:
+                # ignore unreadable directories like import does
+                dircontents = []
+            for fn in dircontents:
+                subname = getmodulename(fn)
+                if subname == '__init__':
+                    ispkg = True
+                    break
+            else:
+                continue    # not a package
+
+        if modname and '.' not in modname:
+            yielded[modname] = 1
+            yield prefix + modname, ispkg
+
+
+# Monkeypatch both path_hooks and iter_importer_modules to make our
```

Comment:
Thanks! I just skimmed the PR quickly, but I think we can avoid monkeypatching iter_importer_modules by implementing an iter_modules method in HyFileFinder.